### PR TITLE
Allow using `Self::` in doc

### LINKED ISSUE #70802: Resolve intra-doc-links on Self for method
I tried this code:

```rust
#![deny(intra_doc_link_resolution_failure)]


/// [Self::bar]
pub struct Works;

impl Works {
    pub fn bar() {}
}

pub struct Broken;

impl Broken {
    /// [Self::bar]
    pub fn bar() {}
}

```

I expected to see this happen:
Docs for `Broken::bar` is generated correctly by running `cargo +nightly doc` 

Instead, this happened:
I'm getting this error
```
> cargo +nightly doc
 Documenting rust-playground v0.1.0 (/home/xliiv/workspace/rust-playground)
error: `[Self::bar]` cannot be resolved, ignoring it.
  --> src/lib.rs:14:10
   |
14 |     /// [Self::bar]
   |          ^^^^^^^^^ cannot be resolved, ignoring
   |
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
1  | #![deny(intra_doc_link_resolution_failure)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`

error: aborting due to previous error

error: Could not document `rust-playground`.

Caused by:
  process didn't exit successfully: `rustdoc --edition=2018 --crate-type lib --crate-name rust_playground src/lib.rs -o /home/xliiv/workspace/rust-playground/target/doc --error-format=json --json=diagnostic-rendered-ansi -L dependency=/home/xliiv/workspace/rust-playground/target/debug/deps` (exit code: 1)
```

### Meta

`rustc --version --verbose`:
```
> rustc --version --verbose
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-unknown-linux-gnu
release: 1.42.0
LLVM version: 9.0

```

### PR DESCRIPTION
**Pull request by ghost**
Closes #70802

### CODE REVIEW

**Comment by Dylan-DPC** [comment]
We do not allow merge commits, kindly rebase and re-commit the change :)

File: src/librustdoc/fold.rs
Comment (lzutao) [inline src/librustdoc/fold.rs:54]: Use `lazy_static!` as https://docs.rs/regex/1.3.7/regex/#example-avoid-compiling-the-same-regex-in-a-loop points out

Comment (xliiv) [reply src/librustdoc/fold.rs]: Good point, thanks!

Comment (xliiv) [reply src/librustdoc/fold.rs]: and done

**Comment by xliiv** [comment]
> We do not allow merge commits, kindly rebase and re-commit the change :)

thank you and done

File: src/librustdoc/fold.rs
Comment (GuillaumeGomez) [inline src/librustdoc/fold.rs:6]: Please use [lazycell](https://docs.rs/lazycell/1.2.1/lazycell/) instead.

File: src/librustdoc/fold.rs
Comment (GuillaumeGomez) [inline src/librustdoc/fold.rs:80]: I'm not sure to understand why we need a regex replacement here... Isn't it possible to just remove the first 6 characters if it starts with "Self::" and prepend with the new text instead? I think this is completely overkill.

File: src/test/rustdoc/intra-link-self.rs
Comment (GuillaumeGomez) [inline src/test/rustdoc/intra-link-self.rs:45]: Why are you returning `Self` everytime btw?

File: src/librustdoc/fold.rs
Comment (xliiv) [reply src/librustdoc/fold.rs:80]: the 6 characters should work in most cases fine.
However, there is one case when it fails, it's `Self` in name, e.g.
```
/// Use MySelf::new to ..
struct MySelf 
```
1. for regex it will ends up like this
```
/// Use MySelf::new to ..
struct MySelf 
```
2. for the 6 replacement ..
```
/// Use MyMySelf::new to ..
struct MySelf 
```
these test cases cover that
https://github.com/rust-lang/rust/pull/71289/files#diff-aefbb1cc61440df9316be7cd6b0f7c2eR66-R82 


On the other hand, i can imagine that using `Self` in the struct name is not common case, so it could be skipped as well. I feel it should be covered so i did.

File: src/test/rustdoc/intra-link-self.rs
Comment (xliiv) [reply src/test/rustdoc/intra-link-self.rs:45]: It's mistake, I copy/pasted the `new` method and forgot to update it.
I'll remove it, thanks!

File: src/librustdoc/fold.rs
Comment (GuillaumeGomez) [reply src/librustdoc/fold.rs:80]: The important part here was:

> if it starts with "Self::"

;)

Comment (xliiv) [reply src/librustdoc/fold.rs:80]: Does it mean it's ok now, are should i change it? :)
If we're changing it, i need more details or ready-to-apply solution ..
Also I'm open to discuss it live somewhere..
I'm on Rust's
1. Discord 
2. Riot

Comment (Dylan-DPC) [reply src/librustdoc/fold.rs:80]: you can ping in #rustdoc channel on official discord server

File: src/librustdoc/fold.rs
Comment (xliiv) [reply src/librustdoc/fold.rs:6]: I'll do it, but I'm just curious, why? I like to know :)

File: src/librustdoc/fold.rs
Comment (GuillaumeGomez) [reply src/librustdoc/fold.rs:80]: When adding new dependencies, one should have good arguments to support it. You're adding both `regex` and `lazy_static`/`lazy_cell` for one thing. And I'm still not sure to understand *why* it's needed. From what I understand, if the string starts with `Self::`, you want to replace it. So why would a regex be needed there?

Comment (xliiv) [reply src/librustdoc/fold.rs:80]: Did you see https://github.com/rust-lang/rust/pull/71289#discussion_r411214750
?

Comment (GuillaumeGomez) [reply src/librustdoc/fold.rs:80]: No it doesn't. If it doesn't start with `Self::`, then you don't apply this. Otherwise I'm missing something but can't see what...

Comment (xliiv) [reply src/librustdoc/fold.rs:80]: We can't speak of "start with" because we need to replace substrings. 
E.g.

```
[my link](Self::Struct)
```

Maybe, please provide your code for the replacement part and i'll use it?

Comment (xliiv) [reply src/librustdoc/fold.rs:80]: On the other hand all possible values are
```
[link](Self::Struct)
[link]: Self::Iterator
[link]:\tSelf::Iterator
[link]:Self::Iterator
```
if my correct.

They could be handled with a ifs, splits and statswith as you suggest.

Comment (GuillaumeGomez) [reply src/librustdoc/fold.rs:80]: The string is really all `[link](Self::Struct)`? If so I'm very surprised... In this case, I thought it was only "Self::Struct". Can you print the actual string at this point please?

Comment (xliiv) [reply src/librustdoc/fold.rs:80]: > Can you print the actual string at this point please?

Which string? :) There can be many string, eg.

```
[link](Self::Struct)
[link]: Self::Iterator
[link]:\tSelf::Iterator
[link]:Self::Iterator
```

Comment (xliiv) [reply src/librustdoc/fold.rs:80]: Also, see
https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md#detailed-design

**Comment by GuillaumeGomez** [comment]
After a discussion with @xliiv, I got a better understanding of why they were using `regex`. It currently runs not on links but on the doc block directly. It brings a few concerns though: working on a text block is definitely not a good idea. For example, if we have:

````rust
/// Text
/// [Link][Self::foo]
///
/// ```
/// // We are inside a trait implementation
/// let x = &array[foo][Self::x];
/// ```
````

`[Link][Self::foo]` will be replaced correctly, but `&array[foo][Self::x];` won't. So we definitely shouldn't do that.

I thought about two solutions to handle this:
 1. Keep the same logic but then use the markdown parser and then replace the links. We just have to rebuild the doc block afterwards but it's not really complicated (we do something equivalent elsewhere). **However**, this solution will break the span and when we'll have warnings, it won't match the actual code (if you replace "Self::" with "SomeSuperLongType::" for example), which is problematic. So I don't think it will work out, or at least not nicely.
 2. Replace `Self::` when we are in the intra-doc link code and give it knowledge of the current type so it can replace `Self::` correctly.

Now that I think about it, when you declare a trait or even a struct, I think that `Self::` should be replaced there as well, meaning `impl` blocks aren't the only ones that should receive this update. Meaning that the second solution looks a lot more promising since in addition to that, it doesn't break spans.

**Comment by xliiv** [comment]
>     2\. Replace `Self::` when we are in the intra-doc link code and give it knowledge of the current type so it can replace `Self::` correctly.

This makes sense more to me, so the solution would be to
1. store somewhere name of `impl` which is available here
https://github.com/rust-lang/rust/blob/25f070d1edf3d54d8933e00eeac95ce55ca8eabb/src/librustdoc/fold.rs#L52
2. use it here
https://github.com/rust-lang/rust/blob/25f070d1edf3d54d8933e00eeac95ce55ca8eabb/src/librustdoc/passes/collect_intra_doc_links.rs#L126

> when you declare a trait or even a struct, I think that Self:: should be replaced there as well,
this just works now AFAIK (see https://github.com/rust-lang/rust/issues/70802)

@GuillaumeGomez if you can hint me where is the right place to store `impl` it could be easier.. now, i'll just check how it works for `struct` block

**Comment by xliiv** [comment]
Btw, this example data
```
inner: ImplItem(
    Impl {
        unsafety: Normal,
        generics: Generics {
            params: [],
            where_predicates: [],
        },
        provided_trait_methods: {},
        trait_: None,
        for_: ResolvedPath {
            path: Path {
                global: false,
                res: Def(
                    Struct,
                    DefId(0:3 ~ foo[8787]::MyStruct[0]),
                ),
                segments: [
                    PathSegment {
                        name: "MyStruct",
                        args: AngleBracketed {
                            args: [],
                            bindings: [],
                        },
                    },
                ],
            },
            param_names: None,
            did: DefId(0:3 ~ foo[8787]::MyStruct[0]),
            is_generic: false,
        },
        items: [
            Item {
                source: Span {
                    filename: Real(
                        "/home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs",
                    ),
                    cnum: crate0,
                    loline: 11,
                    locol: 4,
                    hiline: 13,
                    hicol: 5,
                    original: /home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs:11:5: 13:6,
                },
                name: Some(
                    "resolves_ok",
                ),
                attrs: Attributes {
                    doc_strings: [
                        SugaredDoc(
                            0,
                            /home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs:8:5: 10:47,
                            "[`resolves_ok`]\n\n[`resolves_ok`]: MyStruct::resolves_ok",
                        ),
                    ],
                    other_attrs: [],
                    cfg: None,
                    span: Some(
                        /home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs:8:5: 8:24,
                    ),
                    links: [],
                    inner_docs: false,
                },
                inner: MethodItem(
                    Method {
                        generics: Generics {
                            params: [],
                            where_predicates: [],
                        },
                        decl: FnDecl {
                            inputs: Arguments {
                                values: [],
                            },
                            output: Return(
                                Generic(
                                    "Self",
                                ),
                            ),
                            c_variadic: false,
                            attrs: Attributes {
                                doc_strings: [],
                                other_attrs: [],
                                cfg: None,
                                span: None,
                                links: [],
                                inner_docs: false,
                            },
                        },
                        header: FnHeader {
                            unsafety: Normal,
                            constness: NotConst,
                            asyncness: NotAsync,
                            abi: Rust,
                        },
                        defaultness: Some(
                            Final,
                        ),
                        all_types: [],
                        ret_types: [],
                    },
                ),
                visibility: Public,
                def_id: DefId(0:6 ~ foo[8787]::{{impl}}[0]::resolves_ok[0]),
                stability: None,
                deprecation: None,
            },
            Item {
                source: Span {
                    filename: Real(
                        "/home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs",
                    ),
                    cnum: crate0,
                    loline: 17,
                    locol: 4,
                    hiline: 19,
                    hicol: 5,
                    original: /home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs:17:5: 19:6,
                },
                name: Some(
                    "resolves_bad",
                ),
                attrs: Attributes {
                    doc_strings: [
                        SugaredDoc(
                            0,
                            /home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs:14:5: 16:45,
                            "[`resolves_bad`]\n\n[`resolves_bad`]: Self::resolves_bad",
                        ),
                    ],
                    other_attrs: [],
                    cfg: None,
                    span: Some(
                        /home/xliiv/workspace/rust/src/test/rustdoc/intra-link-self.rs:14:5: 14:25,
                    ),
                    links: [],
                    inner_docs: false,
                },
                inner: MethodItem(
                    Method {
                        generics: Generics {
                            params: [],
                            where_predicates: [],
                        },
                        decl: FnDecl {
                            inputs: Arguments {
                                values: [],
                            },
                            output: Return(
                                Generic(
                                    "Self",
                                ),
                            ),
                            c_variadic: false,
                            attrs: Attributes {
                                doc_strings: [],
                                other_attrs: [],
                                cfg: None,
                                span: None,
                                links: [],
                                inner_docs: false,
                            },
                        },
                        header: FnHeader {
                            unsafety: Normal,
                            constness: NotConst,
                            asyncness: NotAsync,
                            abi: Rust,
                        },
                        defaultness: Some(
                            Final,
                        ),
                        all_types: [],
                        ret_types: [],
                    },
                ),
                visibility: Public,
                def_id: DefId(0:7 ~ foo[8787]::{{impl}}[0]::resolves_bad[0]),
                stability: None,
                deprecation: None,
            },
        ],
```

**Comment by xliiv** [comment]
TODO: check fields too

```
struct Foo {
     /// This is inside [Self::whatever]
     pub f: u8;
}
```

thanks to @GuillaumeGomez

**Comment by Dylan-DPC** [comment]
@xliiv we do not allow merge commits. can you rebase instead? thanks

**Comment by xliiv** [comment]
> @xliiv we do not allow merge commits. can you rebase instead? thanks

It's still WIP (just updated the title), i'll do it once it's done, ok?
Anyway thanks for catching that.

**Comment by Dylan-DPC** [comment]
I understand, but it is easier fixing it now with lesser commits than when you have more commits to rebase :D

**Comment by xliiv** [comment]
Blocked by https://github.com/rust-lang/rust/pull/72173

**Comment by GuillaumeGomez** [comment]
Please ping me once #72173 is merged (and please rebase at that time so the CI can pass too!) so I can review with everything ready. :)

File: src/librustdoc/passes/collect_intra_doc_links.rs
Comment (xliiv) [inline src/librustdoc/passes/collect_intra_doc_links.rs:562]: Could this be written better?

Comment (lzutao) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: Use match and fallback?

**Comment by xliiv** [comment]
> Please ping me once #72173 is merged (and please rebase at that time so the CI can pass too!) so I can review with everything ready. :)

@GuillaumeGomez done and ready for review :)

File: src/librustdoc/passes/collect_intra_doc_links.rs
Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: Could you provide a snippet?

Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: like this?
```rust
                let resolved_self = match (path_str.starts_with("Self::"), parent_name) {
                    (true, Some(ref name)) => {
                        path_str.replace("Self::", &format!("{}::", name))
                    }
                    _ => path_str.to_string()
                }
```

Comment (lzutao) [inline src/librustdoc/passes/collect_intra_doc_links.rs:562]:
```suggestion
                let resolved_self;
                if let Some(remain) = path_str.strip_prefix("Self::") {
                    if let Some(ref name) = parent_name {
                        resolved_self = format!("{}::{}", name, remain),
                        path_str = &resolved_self;              
                    }
                }
```

Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]:
```
error[E0658]: use of unstable library feature 'str_strip': newly added
   --> src/librustdoc/passes/collect_intra_doc_links.rs:554:48
    |
554 |                 if let Some(remain) = path_str.strip_prefix("Self::") {
    |                                                ^^^^^^^^^^^^
    |
    = note: see issue #67302 <https://github.com/rust-lang/rust/issues/67302> for more information
    = help: add `#![feature(str_strip)]` to the crate attributes to enable
```

Comment (lzutao) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: You need to add the feature attr. It is okay to do so in std.

Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: Or maybe
```
                if path_str.starts_with("Self::") {
                    if let Some(ref name) = parent_name {
                        resolved_self = path_str.replace("Self::", &format!("{}::", name));
                        path_str = &resolved_self;
                    }
                }
```
i can do both way so up to you

Comment (lzutao) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: It is not up to me but to authoritative reviewers. I'd use unstable feature because I see it nicer.

Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: Sure. I'm waiting then for a suggestion.

Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: Actually, who do you mean by "authoritative reviewers"?

Comment (lzutao) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: Rustdoc reviewers. Sorry I am not a native speaker.

Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:562]: You mean e.g @GuillaumeGomez ? Nevermind, i'll wait more.


> Sorry I am not a native speaker.

Who cares. I mean, I like both non-native and native speakers. :)

File: src/librustdoc/lib.rs
Comment (GuillaumeGomez) [inline src/librustdoc/lib.rs:16]: What is it used for?

Comment (xliiv) [reply src/librustdoc/lib.rs:16]: enables `if let Some(remain) = path_str.strip_prefix("Self::")`

see thead here https://github.com/rust-lang/rust/pull/71289#discussion_r425792312

File: src/librustdoc/passes/collect_intra_doc_links.rs
Comment (GuillaumeGomez) [inline src/librustdoc/passes/collect_intra_doc_links.rs:566]: Nice trick!

Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs:566]: credits go for @lzutao

**Review by GuillaumeGomez — COMMENTED** [review]
Looks all good to me! Just the newly added feature that seems weird.

**Comment by GuillaumeGomez** [comment]
My question has been answered, therefore we can move forward. Thanks a lot for implementing it!

@bors: r+

File: src/librustdoc/passes/collect_intra_doc_links.rs
Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: used `path_str.strip_prefix("Self::")` so resolving..

**Comment by xliiv** [comment]
> Just the newly added feature that seems weird.

Also it could be replaced with
https://github.com/rust-lang/rust/pull/71289#discussion_r425856612

**Review by GuillaumeGomez — APPROVED** [review]

**Pull request merged**
